driver-async/src/main/com/mongodb/async/client/OperationExecutorImpl.java

```diff
@@ -134,15 +170,92 @@ private void labelException(final Throwable t, final ClientSession session) {
+                    } else {
+                        if (clusterType == ClusterType.SHARDED) {
+                            bindWithPinnedMongos(cluster, readPreference, session, errHandlingCallback, callback);
+                        } else {
+                            allocateReadWriteBinding(readPreference, readConcern, session, ownsSession, callback);
+                        }
+                    }
+                }
+            });
+        } else {
+            allocateReadWriteBinding(readPreference, readConcern, null, ownsSession, callback);
+        }
+    }
+
+    private void bindWithPinnedMongos(final Cluster cluster, final ReadPreference readPreference, final ClientSession session,
+                                      final SingleResultCallback<AsyncReadWriteBinding> errHandlingCallback,
+                                      final SingleResultCallback<AsyncReadWriteBinding> callback) {
+        if (session.getPinnedMongosAddress() == null) {
+            cluster.selectServerAsync(
+                    new ReadPreferenceServerSelector(getReadPreferenceForBinding(readPreference, session)),
+                    new SingleResultCallback<Server>() {
+                        @Override
+                        public void onResult(final Server server, final Throwable t) {
+                            if (t != null) {
+                                errHandlingCallback.onResult(null, t);
+                            } else {
+                                session.setPinnedMongosAddress(server.getDescription().getAddress());
+                                callback.onResult(new AsyncSingleServerBinding(cluster,
+                                        server.getDescription().getAddress()), null);
+                            }
+                        }
+                    });
+        } else {
+            callback.onResult(new AsyncSingleServerBinding(cluster, session.getPinnedMongosAddress()), null);
+        }
+    }
+
+    private void allocateReadWriteBinding(final ReadPreference readPreference, final ReadConcern readConcern,
+                                          final ClientSession session, final boolean ownsSession,
+                                          final SingleResultCallback<AsyncReadWriteBinding> callback) {
         AsyncReadWriteBinding readWriteBinding = new AsyncClusterBinding(mongoClient.getCluster(),
                 getReadPreferenceForBinding(readPreference, session), readConcern);
         if (session != null) {
-            readWriteBinding = new ClientSessionBinding(session, ownsSession, readWriteBinding);
+            callback.onResult(new ClientSessionBinding(session, ownsSession, readWriteBinding), null);
+        } else {
+            callback.onResult(readWriteBinding, null);
+        }
+    }
+
+    private void getClusterDescription(final Cluster cluster, final SingleResultCallback<ClusterType> callback) {
+        ClusterDescription description = cluster.getDescription();
```

Comment:
This should be cluster.getCurrentDescription(), since cluster.getDescription() can block.